File: c/nth-prime/nth_prime.c
#include "nth_prime.h"

uint32_t nth(uint32_t nth_prime) { return sieve(LIMIT, MAX_PRIMES, nth_prime); }

uint32_t sieve(uint32_t limit, size_t max_primes, size_t nth_prime) {
    uint32_t prime = 0;
    uint32_t prime_counter = 0;

    if (nth_prime <= 0) {
        return 0;
    }

    if (nth_prime <= 2) {
        return nth_prime + 1;
    }

    if (limit < 2 || max_primes == 0) {
        return prime_counter;
    }

    uint32_t *primes = calloc(LIMIT, sizeof(uint32_t));

    bool is_prime[limit + 1];
    memset(is_prime, true, sizeof(is_prime));

    for (uint32_t i = 2; i <= limit; i++) {
        if (is_prime[i]) {
            primes[prime_counter++] = i;

            if (prime_counter >= max_primes) {
                break;
            }

            for (uint32_t j = 2; i * j <= limit; j++) {
                is_prime[i * j] = false;
            }
        }
    }

    size_t count = 0;
    for (uint32_t i = 2; i <= limit; i++) {
        if (is_prime[i]) {
            count += 1;
        }

        if (count == nth_prime) {
            prime = i;
            break;
        }
    }

    free(primes);

    return prime;
}

File: ruby/nth-prime/nth_prime.rb
# frozen_string_literal: false

# https://exercism.org/tracks/ruby/exercises/nth-prime
# nth prime exercise
module Prime
  MAX_PRIMES = 1_000_000

  def self.nth(loc)
    raise ArgumentError, 'prime nth count needs to be > 0' if loc.zero?
    return 2 if loc == 1
    return 3 if loc == 2

    sieve(MAX_PRIMES)[loc - 1]
  end

  def self.sieve(limit)
    result = []
    return result if limit < 2

    result.push(2)
    return result if limit == 2

    @numbers = Array.new(limit + 1, true)
    @numbers.each_index { |number| @numbers[number] = false if number.modulo(2).zero? }
    @numbers[0] = @numbers[1] = false
    @numbers[2] = true

    (3..limit).step(2).each_entry do |number|
      is_prime = @numbers[number]
      next unless is_prime

      jdex = 3
      while (mul = number * jdex) <= limit
        index = mul
        @numbers[index] = false
        jdex += 1
      end
    end

    result = []
    @numbers.each_with_index do |is_prime, index|
      result.push(index) if is_prime
    end

    result
  end
end
